main.go

```diff
@@ -0,0 +1,144 @@
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/carwow/umbra/dispatcher"
+	"github.com/carwow/umbra/pool"
+	"github.com/carwow/umbra/reporter"
+	"github.com/go-redis/redis/v8"
+)
+
+const (
+	defaultRedisURL          = "redis://localhost:6379"
+	defaultRedisChannel      = "umbra_channel"
+	defaultRequestBufferSize = 25
+	defaultTimeout           = 5 * time.Second
+	defaultWorkers           = 100
+	defaultReplication       = 1
+)
+
+func init() {
+	log.SetPrefix("[umbra] ")
+	log.SetFlags(log.Lmsgprefix | log.LstdFlags)
+}
+
+// builds and ping the redis connection
+func initRedis(url string) *redis.Client {
+	options, err := redis.ParseURL(url)
+	if err != nil {
+		log.Fatalf("failed to parse redis URL(%v): %v", defaultRedisURL, err)
+	}
+
+	client := redis.NewClient(options)
+	if err := client.Ping(context.TODO()).Err(); err != nil {
+		log.Fatalf("failed to ping redis %v", err)
+	}
+
+	return client
+}
+
+// builds and pings the redis subscription
+func initPubsub(redisClient *redis.Client, channel string) *redis.PubSub {
+	pubsub := redisClient.Subscribe(context.TODO(), defaultRedisChannel)
+
+	_, err := pubsub.Receive(context.TODO())
+	if err != nil {
+		log.Fatal("error setting up subscription!")
+	}
+
+	return pubsub
+}
+
+// waits for SIGTERM and runs draining and teardown of workers
+func waitForInterrupt(pubsub *redis.PubSub, p pool.Pool, redisClient *redis.Client, wg *sync.WaitGroup) {
+	c := make(chan os.Signal)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+
+	<-c // wait for signal
+
+	log.Println("got SIGTERM")
+	log.Println("unsubscribing...")
+	pubsub.Close()
+
+	log.Println("stopping workers...")
+	p.Stop()
+
+	log.Println("closing redis conn...")
+	redisClient.Close()
+
+	wg.Wait()
+}
+
+func report(r reporter.Reporter) {
+	report := r.Report()
+
+	log.Printf(" * Report\n")
+	log.Printf("\tTotal Processed:    %v\n", report.ProcessedTotal)
+	log.Printf("\tTotal Replicated:   %v\n", report.ReplicatedTotal)
+	log.Printf("\tTotal Errored:      %v\n", report.ErroredTotal)
+	log.Printf("\tTotal Server Error: %v\n", report.ServerErrorTotal)
+	log.Printf("\tTotal Server OK:    %v\n", report.ServerOKTotal)
+	log.Printf("\tTotal DurationMs:   %v\n", report.TotalDurationMs)
+}
+
+func main() {
+	redisURL := flag.String("redis", defaultRedisURL, "redis connection string")
+	workers := flag.Int("workers", defaultWorkers, "number of concurrent workers")
+	buffer := flag.Int("buffer", defaultRequestBufferSize, "request buffer size")
+	timeout := flag.Duration("timeout", defaultTimeout, "http client timeout duration")
+	replication := flag.Float64("replication", defaultReplication, "number of times to replicate requests")
+
+	flag.Parse()
+
+	client := initRedis(*redisURL)
+	pubsub := initPubsub(client, defaultRedisChannel)
+	channel := pubsub.ChannelSize(*buffer)
+
+	p := pool.New(*workers, *buffer, *timeout).Start()
+	r := reporter.New(p.ResultsChannel())
+	d := dispatcher.New(p, r, channel, *replication)
```

Comment:
The code is simple enough that we probably don't need to worry about this, but a few main points that jump to mind when reading the code that we could tackle later.

1. Should the pubsub and channel be setup inside the dispatcher?
2. Should the dispatcher and reporter also have Start/Stop functions?
3. The pool doesn't knows about the reporter and pushes results into a channel which the reporter consumes, but the dispatcher knows about the reporter and the pool and just calls functions on both. Should (option 1) the pool also call a function on the reporter (requires moving the Result struct into the reporter), or (option 2) the dispatcher just pushes requests into a channel (without knowing about the pool)? Option 2 is interesting for decoupling the components, but option 1 is the more tradition approach on other languages and may make the flow of data more easy to follow. 🤔 
4. Should the report function in main be inside the reporter? (Specially if the reporter has Start function). We could then have different implementations of a reporter, e.g. ConsoleReporter, WebReporter (this one could start a web server and show the stats in a webpage for example). I think here I am also being influenced by the name, we probably have two interfaces, e.g. one would be responsible for displaying the stats (Reporter), and one would be responsible for collecting the stats (ResultsCollector). 🤔 At the moment, the second one is called Reporter, but not sure if it is the best name. The first one is just implemented in the main file.
 
No need to make any changes, just food for thought. The overall code looks really good 👍

Reply:
1. 👍 
2. this is a good idea, having these would allow for more fine control over the consumption behaviour, atm it is reliant on the channel closing for consumption to stop. if we add Start/Stop we can expose for fine-grain control.
3. I think I prefer option 1, I think it is too much responsibility for the `pool` to be managing the `resultsChan` actually. It makes more sense for that to be a detail within the `reporter`.
4. I agree here that there is a mix of concerns, separating into a collector and reporter makes sense to me 👍 

I will create an issue to capture these changes, I think it will be nicer to make them with a smaller diff to look at 😄